### ISSUE TITLE
Errors During Instrumentation (JMS and others)

### ISSUE THREAD

**Issue by tony-schellenberg** [DataDog/dd-trace-java#517]
Using version 0.15.0 with debug tracing enabled I see quite a few exceptions similar to the following:
```
[main] DEBUG datadog.trace.agent.tooling.ByteBuddyElementMatchers - Instrumentation type matcher unexpected exception: datadog.trace.instrumentation.jms.JMSMessageProducerInstrumentationjava.lang.IllegalStateException: Cannot resolve type description for net.bytebuddy.dynamic.Nexus
    at net.bytebuddy.pool.TypePool$Resolution$Illegal.resolve(TypePool.java:135)
    at net.bytebuddy.pool.TypePool$Default$WithLazyResolution$LazyTypeDescription.delegate(TypePool.java:1252)
    at net.bytebuddy.description.type.TypeDescription$AbstractBase$OfSimpleType$WithDelegation.getModifiers(TypeDescription.java:7121)
    at net.bytebuddy.matcher.ModifierMatcher.matches(ModifierMatcher.java:31)
    at net.bytebuddy.matcher.ModifierMatcher.matches(ModifierMatcher.java:12)
    at net.bytebuddy.matcher.NegatingMatcher.matches(NegatingMatcher.java:29)
    at net.bytebuddy.matcher.ElementMatcher$Junction$Conjunction.matches(ElementMatcher.java:101)
    at datadog.trace.agent.tooling.ByteBuddyElementMatchers$SafeMatcher.matches(ByteBuddyElementMatchers.java:262)
    at net.bytebuddy.agent.builder.AgentBuilder$RawMatcher$ForElementMatchers.matches(AgentBuilder.java:1222)
    at net.bytebuddy.agent.builder.AgentBuilder$RawMatcher$Conjunction.matches(AgentBuilder.java:1079)
    at net.bytebuddy.agent.builder.AgentBuilder$Default$Transformation$Simple.matches(AgentBuilder.java:9033)
    at net.bytebuddy.agent.builder.AgentBuilder$Default$Transformation$Compound.matches(AgentBuilder.java:9267)
    at net.bytebuddy.agent.builder.AgentBuilder$RedefinitionStrategy$Collector.consider(AgentBuilder.java:6186)
    at net.bytebuddy.agent.builder.AgentBuilder$RedefinitionStrategy.apply(AgentBuilder.java:4383)
    at net.bytebuddy.agent.builder.AgentBuilder$Default.installOn(AgentBuilder.java:8523)
    at net.bytebuddy.agent.builder.AgentBuilder$Default$Delegator.installOn(AgentBuilder.java:10182)
    at datadog.trace.agent.tooling.AgentInstaller.installBytebuddyAgent(AgentInstaller.java:92)
    at datadog.trace.agent.tooling.AgentInstaller.installBytebuddyAgent(AgentInstaller.java:32)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at datadog.trace.agent.TracingAgent.startAgent(TracingAgent.java:72)
    at datadog.trace.agent.TracingAgent.premain(TracingAgent.java:37)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:386)
    at sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:401)
```

In my particular case, I'm trying to get JMS Producer instrumentation working but I'm not seeing any spans created for it. I am getting my JaxRS server spans, HttpURLConnection, etc but not JMS.

I'm guessing that something is failing with the instrumentation and that's why it's not working. Is that a reasonable guess?

**Comment by realark** [comment]
@tony-schellenberg I don't think failing to resolve `net.bytebuddy.dynamic.Nexus` would affect the JMS instrumentation. I think in this case that's not actually an error due to the class being dynamically generated.

I can take a look through your full debug log if you can send those.

**Comment by tony-schellenberg** [comment]
I think I can do that. Do you just need all the DataDog lines?

**Comment by realark** [comment]
Yes, that will do it. Thanks!

**Comment by tony-schellenberg** [comment]
I'm just looking at the logs. Is there a secure way for me to transfer them to you? I don't think there's anything sensitive in them but it's probably better. Should I open a ticket with support and provide them there and give you the ticket number here?

**Comment by realark** [comment]
Yeah going through the support system and posting the ticket number is the best way to get those logs to me.

**Comment by tony-schellenberg** [comment]
I have posted a log to ticket #171383

**Comment by realark** [comment]
> I have posted a log to ticket #171383

I took a look through the logs. I think the missing `Nexus` class is a red herring. I don't see the JMS Instrumentation attempting to apply. Which JMS client (and version) are you running with?

**Comment by tylerbenson** [comment]
@tony-schellenberg Is that even available in maven central?  I only see version 4.2+ [there](https://mvnrepository.com/artifact/com.sun.messaging.mq/imq).
We'd be happy to run it through some tests to verify compatibility, but that becomes a challenge if it's not in maven central.

**Comment by tony-schellenberg** [comment]
I don't know if it's there or not. We had the product since before it was OpenMQ back when it was paid. We have our own Maven server that mirrors maven central but also has other artifacts so this one might not be generally available. 

Given that it supports the javax.jms interfaces, can you think of a reason that it wouldn't instrument? Is there something that I can look for in it or a way that I can run the tests?

**Comment by tylerbenson** [comment]
@tony-schellenberg When running with debug logging enabled, I'd expect to see these kinds of messages:
```
[Test worker] DEBUG datadog.trace.agent.tooling.Instrumenter$Default - Applying instrumentation: okhttp -- datadog.trace.instrumentation.okhttp3.OkHttp3Instrumentation on sun.misc.Launcher$AppClassLoader@15db9742
[Test worker] DEBUG datadog.trace.agent.tooling.AgentInstaller$LoggingListener - Transformed class okhttp3.OkHttpClient -- sun.misc.Launcher$AppClassLoader@15db9742
[Test worker] DEBUG datadog.trace.bootstrap.ExceptionLogger - Failed to handle exception in instrumentation
```

Another message you should search for in the logs is `Instrumentation muzzled`, which might give indication as to why the library might be skipped.

**Comment by tony-schellenberg** [comment]
I can't see anything in the logs referring to instrumentation being muzzled.

Looking at this https://github.com/DataDog/dd-trace-java/blob/master/dd-java-agent/instrumentation/jms/src/main/java/datadog/trace/instrumentation/jms/JMSMessageProducerInstrumentation.java#L41

I unpacked the imq library and did a bit of digging. There is a concrete implementation that implements javax.jms.MessageProducer. I'm don't have much understanding of how the auto instrumentation works. Is there some other obvious reason why it wouldn't be able to find it that I can dig into?

**Comment by realark** [comment]
Generally the way our instrumentation works:

1. A matcher sees a class it could potentially instrument (like in your jms-producer link)
2. The classpath is checked to ensure it's safe to apply instrumentation.
3. Instrumentation applies or logs a failure to apply.

Any of these stages would have produced something in the log file.

I think the issue is we're skipping instrumentation because the class name starts with `com.sun`. We skip "core" namespaces for performance reasons.

https://github.com/datadog/dd-trace-java/blob/master/dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentInstaller.java#L70

Can you confirm the package-namespace of the client you're trying to instrument?

**Comment by tony-schellenberg** [comment]
I think you might have figured it out. The one class I'm looking at so far is com.sun.messaging.jmq.jmsclient.MessageProducerImpl

**Comment by realark** [comment]
Great, that explains it. We should be good to whitelist `com.sun.messaging.`. Do you think that will cover your client's namespace?

**Comment by tony-schellenberg** [comment]
From what I can see of it, I think that should be good. Thanks for helping me out with this.

**Comment by tylerbenson** [comment]
Fix is included in [0.17.0](https://github.com/DataDog/dd-trace-java/releases/tag/v0.17.0).
